sql/index.go

```diff
@@ -215,61 +222,109 @@ func (r *IndexRegistry) RegisterIndexDriver(driver IndexDriver) {
 	r.drivers[driver.ID()] = driver
 }
 
-// LoadIndexes loads all indexes for all dbs, tables and drivers.
+// LoadIndexes creates load functions for all indexes for all dbs, tables and drivers.  These functions are called
+// as needed by the query
 func (r *IndexRegistry) LoadIndexes(dbs Databases) error {
 	r.driversMut.RLock()
 	defer r.driversMut.RUnlock()
 	r.mut.Lock()
 	defer r.mut.Unlock()
 
-	for _, driver := range r.drivers {
-		for _, db := range dbs {
-			for _, t := range db.Tables() {
-				indexes, err := driver.LoadAll(db.Name(), t.Name())
-				if err != nil {
-					return err
-				}
+	ctx := context.TODO()
+	for drIdx := range r.drivers {
+		driver := r.drivers[drIdx]
+		for dbIdx := range dbs {
+			db := dbs[dbIdx]
+			tNames, err := db.GetTableNames(ctx)
+
+			if err != nil {
+				return err
+			}
+
+			for tIdx := range tNames {
+				tName := tNames[tIdx]
+
+				loadF := func(ctx context.Context) error {
+					t, ok, err := db.GetTableInsensitive(ctx, tName)
```

Comment:
You need to deal with the lock semantics in this function scope instead of in the parent scope

Reply:
discussed